### ISSUE TITLE
ability to expose metrics from multiple integrations on a /metrics endpoint

### ISSUE THREAD

**Issue by rohankmr414** [grafana/alloy#576]
### Request

I want to use grafana alloy for exposing node exporter and process exporter metrics, in the documentation I can see that there's a way to forward those metrics to a prometheus remote write url, I wanted to know if there's a way to expose those metrics on a /metrics type endpoint in grafana alloy.

### Use case

We want to use it in our already existing monitoring system which scrapes metrics from a huge fleet of EC2 instances, we don't want to remote write from every EC2 instance.

**Comment by tpaschalis** [comment]
The answer is.. this kinda exists today already.

In Alloy's built-in debug UI, you can see the endpoint that each exporter component exposes internally.

![image](https://github.com/grafana/alloy/assets/17771679/1f7481bd-cb71-4fe1-8ed5-cf39e0b50f26)

The inter-component communication happens over a memory connection as an optimization, but you can hit the same endpoint to get the metrics (in this case `localhost:12345/api/v0/component/prometheus.exporter.self.integrations_alloy/metrics`) will give you all the

```
$ curl localhost:12345/api/v0/component/prometheus.exporter.self.integrations_alloy/metrics
# HELP alloy_build_info A metric with a constant '1' value labeled by version, revision, branch, goversion from which alloy was built, and the goos and goarch for the build.
# TYPE alloy_build_info gauge
alloy_build_info{branch="HEAD",goarch="amd64",goos="linux",goversion="go1.22.1",revision="1eef9b3ae",tags="netgo,builtinassets,promtail_journal_enabled",version="v1.0.0"} 1
...
```

Given this is a side-effect of how inter-component communication happens and is only available via a an internal `/api/v0/` route, we may want to find a different/more robust way to achieve this in the future.

**Comment by rohankmr414** [comment]
@tpaschalis I was looking for an option to configure a single endpoint for serving multiple integrations, just like in [fluent bit](https://docs.fluentbit.io/manual/pipeline/outputs/prometheus-exporter) where you can configure what needs to be exposed. @rfratto mentioned in the [community slack](https://grafana.slack.com/archives/C01050C3D8F/p1713182616490979?thread_ts=1713176044.922109&cid=C01050C3D8F) that the stability of this API is not guaranteed.

**Comment by tpaschalis** [comment]
Yeah, you're right, I see, thanks. I've amended the comment above.

**Comment by itjobs-levi** [comment]
Hi!

There may be several ways,
but in my case, I do not prefer that the collection target server directly puts it into the metric storage.
This is because the configuration can change frequently and the collection target server does not have enough resources to configure the Alloy agent for high availability. Also, if collection target server filtering is required, it can be managed centrally from the collector cluster.

(Usually, one Alloy agent will be used for the collection target server.)

Therefore, the collection target server will only perform up to prometheus.*.exporter
and configure the collector as a cluster to scrape from multiple servers.

I am scraping by specifying /api/v0/comComponent/prometheus.exporter.[unix|process|...].LABEL/metrics.
Of course, it works well.

However, it is said that this is not a recommended method.
If that API is not official, I think that alloy should declare the exporter information exposed by the collection target server when scraping and scrape it from the API. Alternatively, multiple exporter metrics should be added to the /metrics base metrics address.

**Comment by Upanshu11** [comment]
Any update on this? I am already using Grafana Agent and was looking to migrate to Alloy. On top of it, I also have a use case of HPA, where I want a metrics API to expose by custom metrics for autoscaling. 
If alloy exposes my custom metrics on /metrics endpoint, it can help me in avoiding setting up a Prom adapter or other metrics API.

**Comment by Napsty** [comment]
Just stumbled on this after having tried to retrieve node exporter metrics using http://localhost:12345/metrics - where I expected the enabled exporter metrics would show up. 

Thanks for https://github.com/grafana/alloy/issues/576#issuecomment-2056851611, this helps a lot.

**Comment by angelakis** [comment]
Hey I'm also relying on the undocumented API and I'm wondering if there's any plan to stabilize it or offer a different way to scrape metrics from Alloy.

**Comment by Napsty** [comment]
@angelakis To my understanding, Alloy is not supposed to be used in a classical "Exporter" way. 
A Prometheus exporter collects data and presents these as metrics under a certain port and path, making it available for scraping by an external tool (Prometheus or Otel Collector for example).

Alloy does the data/metric collection but does not present these metrics. The metrics are supposed to be forwarded/sent to an endpoint using the `prometheus.writer` component.

I've used Alloy's internal component API just to verify the embedded exporter configuration (`prometheus.exporter.XYZ`) and see if the metrics are actually collected (https://www.claudiokuenzler.com/blog/1462/how-to-scrape-node-exporter-metrics-grafana-alloy). It's very helpful, indeed, but is not supposed to be scraped by other tools/scrapers.
I might be wrong, but that's how I understand what Alloy was built for.
